### PR TITLE
Update autodoc to support stabilizing limbs

### PR DESCRIPTION
**Pull request by psykzz**
## About The Pull Request

Updates the `datum/component/suit_autodoc`.

Updates how cooldowns are handled to remove world.time checks and instead use timers
Adds support for stabalizing limbs.
Removes the need for snowflake `clothing_procs.dm`
Additionally adds documentation

## Why It's Good For The Game

This adds future support for modular armor outside of just the B18 armor.

## Changelog
:cl:
refactor: update the autodoc to support limb stabalizing directly
/:cl:

### CODE REVIEW

File: code/datums/components/suit_autodoc.dm [hunk 1]
```diff
@@ -62,10 +85,16 @@
 	src.brute_chems = brute_chems || default_brute_chems
 	src.tox_chems = tox_chems || default_tox_chems
 	src.pain_chems = pain_chems || default_pain_chems
-	
+
 	if(!isnull(overdose_threshold_mod))
 		src.overdose_threshold_mod = overdose_threshold_mod
 
+	if(!isnull(supporting_limbs))
```
Comment (spookydonut) [inline code/datums/components/suit_autodoc.dm:92]: supporting limbs is a feature of hardsuits in other codebases, i feel like it shouldnt be a feature of the suit autodoc component

File: code/modules/clothing/suits/marine_armor.dm [hunk 2]
```diff
@@ -246,7 +246,7 @@
 
 /obj/item/clothing/suit/storage/marine/specialist/Initialize(mapload, ...)
 	. = ..()
-	AddComponent(/datum/component/suit_autodoc)
+	AddComponent(/datum/component/suit_autodoc, supporting_limbs=src.supporting_limbs)
```
Comment (spookydonut) [inline code/modules/clothing/suits/marine_armor.dm:249]: you cant use named args with addcomponent

**Comment by psykzz** [comment]
I'll rework this into another component then that will handle the limb support

**Comment by spookydonut** [comment]
> 
> 
> I'll rework this into another component then that will handle the limb support

you could do it with a bespoke element

File: code/modules/organs/limbs.dm [hunk 3]
```diff
@@ -938,17 +938,9 @@ Note that amputating the affected organ does in fact remove the infection from t
 	if (prob(25))
 		release_restraints()
 
-	// This is mostly for the ninja suit to stop ninja being so crippled by breaks.
-	// TODO: consider moving this to a suit proc or process() or something during
-	// hardsuit rewrite.
-	if(ishuman(owner))
+	/// Emit a signal for autodoc to support the life if available
+	SEND_SIGNAL(owner, COMSIG_HUMAN_LIMB_FRACTURED, src)
```
Comment (spookydonut) [inline code/modules/organs/limbs.dm:942]: dont need src here first argument of the signal is already the limb as source

File: code/datums/elements/limb_support.dm [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+	if(!ishuman(equipper))
+		return
+	RegisterSignal(equipper, COMSIG_HUMAN_LIMB_FRACTURED, .proc/fuckibrokemyleg)
+	fuckibrokemyleg(equipper, null, FALSE) // manually called when equipped to stabilize limbs
+
+/datum/element/limb_support/proc/dropped(datum/source, mob/equipper)
+	fuckibrokemyleg(equipper, null, TRUE) // manually called when dropped to remove stabilization of limbs
+	UnregisterSignal(equipper, COMSIG_HUMAN_LIMB_FRACTURED)
+
+/datum/element/limb_support/proc/fuckibrokemyleg(datum/owner, datum/limb/fractured_limb, dropped = FALSE)
```
Comment (spookydonut) [inline code/datums/elements/limb_support.dm:25]: i love how you gave it a proper name but then kept this shitpost proc name i made :P

File: code/modules/clothing/suits/marine_armor.dm [hunk 5]
```diff
@@ -241,13 +241,13 @@
...
 	slowdown = SLOWDOWN_ARMOR_VERY_HEAVY
-	supporting_limbs = list(CHEST, GROIN, ARM_LEFT, ARM_RIGHT, HAND_LEFT, HAND_RIGHT, LEG_LEFT, LEG_RIGHT, FOOT_LEFT, FOOT_RIGHT) //B18 effectively stabilizes these.
+	supporting_limbs = CHEST | GROIN | ARM_LEFT | ARM_RIGHT | HAND_LEFT | HAND_RIGHT | LEG_LEFT | LEG_RIGHT | FOOT_LEFT | FOOT_RIGHT //B18 effectively stabilizes these.
 	resistance_flags = UNACIDABLE
 
 /obj/item/clothing/suit/storage/marine/specialist/Initialize(mapload, ...)
 	. = ..()
 	AddComponent(/datum/component/suit_autodoc)
-	AddComponent(/datum/component/suit_limb_support, supporting_limbs)
+	AddElement(/datum/element/limb_support)
```
Comment (spookydonut) [inline code/modules/clothing/suits/marine_armor.dm]:
```suggestion
	AddElement(/datum/element/limb_support, supporting_limbs)
```

**Comment by spookydonut** [comment]
@psykzz reee self merge

**Pull request merged**